src/mds/MDCache.cc

```diff
@@ -13419,6 +13504,286 @@ void MDCache::clear_dirty_bits_for_stray(CInode* diri) {
+    lov.add_rdlock(&in->nestlock);
+    lov.add_rdlock(&in->policylock);
+    // N.B.: NO xlock/wrlock on quiescelock; we need to allow access to mksnap/lookup
+    // This is an unfortunate inconsistency. It may be possible to circumvent
+    // this issue by having those ops acquire the quiscelock only if necessary.
+    if (is_root) {
+      lov.add_rdlock(&in->quiescelock);
+    } else {
+      lov.add_xlock(&in->quiescelock); /* !! */
+    }
+    lov.add_rdlock(&in->snaplock);
+    lov.add_rdlock(&in->xattrlock);
+    if (!mds->locker->acquire_locks(mdr, lov, nullptr, {in}, false, true)) {
+      return;
+    }
+  } else if (!splitauth) {
+    dout(5) << "auth is split and splitauth is false: " << *in << dendl;
+    qs.add_failed(mdr, -CEPHFS_EPERM);
+    mds->server->respond_to_request(mdr, -CEPHFS_EPERM);
+    return;
+  }
+
+  if (in->get_projected_inode()->get_quiesce_block()) {
+    dout(10) << __func__ << " quiesce is blocked for this inode; dropping locks!" << dendl;
+    mdr->mark_event("quiesce blocked");
+    mds->locker->drop_locks(mdr.get());
+    /* keep authpins! */
+    qs.inc_inodes_blocked();
+    mdr->internal_op_finish->complete(0);
+    mdr->internal_op_finish = nullptr;
+    return;
+  }
+
+  if (in->is_dir()) {
+    for (auto& dir : in->get_dirfrags()) {
+      if (!dir->is_auth() && !splitauth) {
+        dout(5) << "auth is split and splitauth is false: " << *dir << dendl;
+        qs.add_failed(mdr, -CEPHFS_EPERM);
+        mds->server->respond_to_request(mdr, -CEPHFS_EPERM);
+        return;
+      }
+    }
+    uint64_t count = 0;
+    MDSGatherBuilder gather(g_ceph_context, new C_MDS_RetryRequest(this, mdr));
+    auto& qops = qrmdr->more()->quiesce_ops;
+    for (auto& dir : in->get_dirfrags()) {
+      for (auto& [dnk, dn] : *dir) {
+        auto* in = dn->get_projected_inode();
+        if (!in) {
+          continue;
+        }
+
+        if (auto it = qops.find(in); it != qops.end()) {
+          dout(25) << __func__ << ": existing quiesce metareqid: "  << it->second << dendl;
+          if (auto reqit = active_requests.find(it->second); reqit != active_requests.end()) {
+            auto& qimdr = reqit->second;
+            dout(25) << __func__ << ": found in-progress " << qimdr << dendl;
+            continue;
+          }
+        }
+        dout(10) << __func__ << ": scheduling op to quiesce " << *in << dendl;
+
+        MDRequestRef qimdr = request_start_internal(CEPH_MDS_OP_QUIESCE_INODE);
+        qimdr->set_filepath(filepath(in->ino()));
+        qimdr->internal_op_finish = gather.new_sub();
+        qimdr->internal_op_private = qfinisher;
+        qops[in] = qimdr->reqid;
+        qs.inc_inodes();
+        if (delay > 0ms) {
+          mds->timer.add_event_after(delay, new LambdaContext([cache=this,qimdr](int r) {
+            cache->dispatch_request(qimdr);
+          }));
+        } else {
+          dispatch_request(qimdr);
+        }
+        if (!(++count % mds->heartbeat_reset_grace())) {
```

Comment:
This counter should be in the `qs` object, shared by all requests. Otherwise, it won't ever fire for a huge binary tree-like dir structure